test/test_torch.py

```diff
@@ -2905,6 +2905,56 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
             with self.assertWarnsOnceRegex(UserWarning, "torch.use_deterministic_algorithms is in beta"):
                 torch.use_deterministic_algorithms(True)
 
+    # This test checks that each nondeterministic function throws an error as
+    # described in the documentation for torch.use_deterministic_algorithms
+    def test_use_deterministic_algorithms_errors(self, device):
+
+        # Add new test cases to this list any time a nondeterministic error is
+        # added to an operation.
+        #
+        # name                  Name of test case
+        #
+        # test_func             Forward pass function that exercises the
+        #                       nondeterministic operation
+        #
+        # error_device          Device that should throw the error.
+        #                       Set to None if all devices should throw an error.
+        #
+        # caller_name_forward   Name of function shown in error for forward pass.
+        #                       Set to None if the forward should not throw an error.
+        #
+        # caller_name_backward  Name of function shown in error for backward pass.
+        #                       Set to None if the backward should not throw an error.
+        nondeterministic_operations = [
+            # name, test_func, error_device, caller_name_forward, caller_name_backward
+            (
+                'kthvalue',
+                (lambda slf, device: torch.randn(10, device=device).kthvalue(4)),
+                'cuda', 'kthvalue CUDA', None
+            ),
+            (
+                'index_select_backward',
+                (lambda slf, device: torch.randn(10, device=device, requires_grad=True).index_select(0, torch.tensor([3], device=device))),
+                'cuda', None, 'index_add_cuda_'
+            ),
+            (
+                'index_copy',
+                (lambda slf, device: torch.randn(10, device=device).index_copy(0, torch.tensor([3], device=device), torch.randn(1, device=device))),
+                None, 'index_copy', None
+            ),
+        ]
+
+        for name, test_func, error_device, caller_name_forward, caller_name_backward in nondeterministic_operations:
```

Comment:
Perhaps instead of looping through the test cases within a unit test, we could instead do something similar to `method_tests` and `OpInfo`, where we create a full unit test for each of these cases. That would make it easier to debug failures.